File: functions/runtime/src/Perper.WebJobs.Extensions/Model/Stream.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Runtime.CompilerServices;
using System.Threading;
using System.Threading.Tasks;

using Apache.Ignite.Core.Client;
using Apache.Ignite.Linq;

using Microsoft.Extensions.Logging;

using Perper.WebJobs.Extensions.Cache;
using Perper.WebJobs.Extensions.Cache.Notifications;
using Perper.WebJobs.Extensions.Services;

namespace Perper.WebJobs.Extensions.Model
{
    [PerperData(Name = "PerperStream")]
    [System.Diagnostics.CodeAnalysis.SuppressMessage("Naming", "CA1711:Identifiers should not have incorrect suffix", Justification = "<Pending>")]
    public class Stream : IStream
    {
        public string StreamName { get; protected set; }

        [NonSerialized] protected readonly FabricService _fabric;
        [NonSerialized] protected readonly IIgniteClient _ignite;

#pragma warning disable CS8618
        [PerperInject]
        protected Stream(FabricService fabric, IIgniteClient ignite)
        {
            _fabric = fabric;
            _ignite = ignite;
        }
#pragma warning restore CS8618

        public Stream(string streamName, FabricService fabric, IIgniteClient ignite)
            : this(fabric, ignite) => StreamName = streamName;
    }

    [PerperData(Name = "PerperStream")]
    [System.Diagnostics.CodeAnalysis.SuppressMessage("Naming", "CA1711:Identifiers should not have incorrect suffix", Justification = "<Pending>")]
    public class Stream<T> : Stream, IStream<T>
    {
        [NonSerialized] public string? FunctionName; // HACK: Used for Declare/InitiaizeStream

        [NonSerialized] private readonly PerperInstanceData _instance;
        [NonSerialized] private readonly int _parameterIndex;
        [NonSerialized] private readonly PerperBinarySerializer _serializer;
        [NonSerialized] private readonly IState _state;
        [NonSerialized] private readonly ILogger _logger;

        [PerperInject]
        public Stream(PerperInstanceData instance, FabricService fabric, IIgniteClient ignite, PerperBinarySerializer serializer, IState state, ILogger logger)
            : base(fabric, ignite)
        {
            _instance = instance;
            _parameterIndex = instance.GetStreamParameterIndex();
            _serializer = serializer;
            _state = state;
            _logger = logger;
        }

        public Stream(string streamName, PerperInstanceData instance, FabricService fabric, IIgniteClient ignite, PerperBinarySerializer serializer, IState state, ILogger logger)
            : this(instance, fabric, ignite, serializer, state, logger) => StreamName = streamName;

        private StreamAsyncEnumerable GetEnumerable(Dictionary<string, object?> filter, bool replay, bool localToData) => new StreamAsyncEnumerable(this, filter, replay, localToData);

        public IAsyncEnumerator<T> GetAsyncEnumerator(CancellationToken cancellationToken = default) => GetEnumerable(new Dictionary<string, object?>(), false, false).GetAsyncEnumerator(cancellationToken);

        public IAsyncEnumerable<T> DataLocal() => GetEnumerable(new Dictionary<string, object?>(), false, true);

        public IAsyncEnumerable<T> Filter(Expression<Func<T, bool>> filter, bool dataLocal = false) => GetEnumerable(FilterUtils.ConvertFilter(filter), false, dataLocal);

        public IAsyncEnumerable<T> Replay(bool dataLocal = false) => GetEnumerable(new Dictionary<string, object?>(), true, dataLocal);

        public IAsyncEnumerable<T> Replay(Expression<Func<T, bool>> filter, bool dataLocal = false) => GetEnumerable(FilterUtils.ConvertFilter(filter), true, dataLocal);

        public async IAsyncEnumerable<TResult> Query<TResult>(Func<IQueryable<T>, IQueryable<TResult>> query)
        {
            var cache = _ignite.GetCache<long, T>(StreamName).WithKeepBinary<long, T>(); // NOTE: Will not work with forwarding
            var queryable = query(cache.AsCacheQueryable().Select(pair => pair.Value)).Select(x => (object?)x);

            using var enumerator = queryable.GetEnumerator();
            while (await Task.Run(enumerator.MoveNext).ConfigureAwait(false)) // Blocking, should run in background
            {
                yield return (TResult)_serializer.DeserializeRoot(enumerator.Current!, typeof(TResult))!;
            }
        }

        private class StreamAsyncEnumerable : IAsyncEnumerable<T>
        {
            protected Stream<T> Stream { get; }

            public Dictionary<string, object?> Filter { get; private set; }
            public bool Replay { get; private set; }
            public bool LocalToData { get; private set; }

            public StreamAsyncEnumerable(Stream<T> stream, Dictionary<string, object?> filter, bool replay, bool localToData)
            {
                Stream = stream;
                Replay = replay;
                Filter = filter;
                LocalToData = localToData;
            }

            public IAsyncEnumerator<T> GetAsyncEnumerator(CancellationToken cancellationToken = new CancellationToken()) => Impl(cancellationToken).GetAsyncEnumerator(cancellationToken);

            private async IAsyncEnumerable<T> Impl([EnumeratorCancellation] CancellationToken cancellationToken = default)
            {
                try
                {
                    await AddListenerAsync().ConfigureAwait(false);

                    await foreach (var (key, notification) in Stream._fabric.GetNotifications(Stream._instance.InstanceName, Stream._parameterIndex, cancellationToken))
                    {
                        if (notification is StreamItemNotification si)
                        {
                            var cache = Stream._ignite.GetCache<long, object>(si.Cache).WithKeepBinary<long, object>();

                            object value;
                            try
                            {
                                value = await cache.GetAsync(si.Key).ConfigureAwait(false);
                            }
                            catch (KeyNotFoundException)
                            {
                                Stream._logger.LogError($"Error AffinityKey({key}) Cache Not Found key: {si.Key} in {si.Cache}");
                                continue;
                            }


                            await ((State)Stream._state).LoadStateEntries().ConfigureAwait(false);

                            try
                            {
                                yield return (T)Stream._serializer.DeserializeRoot(value, typeof(T))!;
                            }
                            finally
                            {
                                await ((State)Stream._state).StoreStateEntries().ConfigureAwait(false);

                                await Stream._fabric.ConsumeNotification(key).ConfigureAwait(false);
                            }
                        }
                    }
                }
                finally
                {
                    if (Stream._parameterIndex < 0)
                    {
                        await RemoveListenerAsync().ConfigureAwait(false);
                    }
                }
            }

            private async Task AddListenerAsync()
            {
                var streamsCache = Stream._ignite.GetCache<string, StreamData>("streams");
                var currentValue = await streamsCache.GetAsync(Stream.StreamName).ConfigureAwait(false);

                currentValue.Listeners.Add(new StreamListener
                {
                    AgentDelegate = Stream._fabric.AgentDelegate,
                    Stream = Stream._instance.InstanceName,
                    Parameter = Stream._parameterIndex,
                    Filter = Filter,
                    Replay = Replay,
                    LocalToData = LocalToData,
                });

                await streamsCache.PutAsync(Stream.StreamName, currentValue).ConfigureAwait(false);
            }

            private async Task RemoveListenerAsync()
            {
                var streamsCache = Stream._ignite.GetCache<string, StreamData>("streams");
                var currentValue = await streamsCache.GetAsync(Stream.StreamName).ConfigureAwait(false);

                currentValue.Listeners.RemoveAt(currentValue.Listeners.FindIndex(listener =>
                    listener.Stream == Stream._instance.InstanceName && listener.Parameter == Stream._parameterIndex
                ));

                await streamsCache.PutAsync(Stream.StreamName, currentValue).ConfigureAwait(false);
            }
        }
    }
}

File: functions/python/perper/model/stream.py
from perper.model.filter_utils import FilterUtils
from perper.cache.stream_listener import StreamListener
from pyignite import GenericObjectMeta
from collections import OrderedDict
from pyignite.datatypes import *
from perper.cache.notifications import StreamItemNotification


class Stream(
    metaclass=GenericObjectMeta,
    type_name="PerperStream`1[[SimpleData]]",
    schema=OrderedDict([("streamname", String)]),
):
    def set_parameters(self, ignite, fabric, **kwargs):
        self.fabric = fabric
        self.ignite = ignite
        self._ignite = ignite
        self.stream_name = self.streamname

        if len(kwargs) > 0:
            self.set_additional_parameters(kwargs)

        self.function_name = None

    def set_additional_parameters(self, kwargs):
        self.serializer = kwargs["serializer"]
        self.state = kwargs["state"]
        self.instance = kwargs["instance"]
        self.parameter_index = self.instance.get_stream_parameter_index()

    def get_enumerable(self, _filter, replay, local_to_data):
        enumerable = StreamEnumerable(self, _filter, replay, local_to_data)
        return enumerable

    async def get_async_generator(
        self,
    ):
        return self.get_enumerable({}, False, False).async_generator()

    def data_local(
        self,
    ):
        return self.get_enumerable({}, False, True)

    def _filter(self, _filter, data_local=False):
        return self.get_enumerable(
            FilterUtils.convert_filter(_filter), False, data_local
        )

    def replay(self, data_local=False, **kwargs):
        if len(kwargs > 0):
            return self.get_enumerable(
                FilterUtils.convert_filter(kwargs["filter"]), True, data_local
            )
        else:
            return self.get_enumerable({}, True, data_local)

    def query(self, query):
        cache = self.ignite.get_cache(self.stream_name)
        queryable = query([pair.value for pair in cache.scan()])

        for (index, value) in enumerate(queryable):
            yield self.serializer.deserialize(value)


class StreamEnumerable:
    def __init__(self, stream, _filter, local_to_data, replay):
        self._stream = stream
        self.local_to_data = local_to_data
        self._filter = _filter
        self.replay = replay

    async def async_generator(
        self,
    ):
        try:
            self.__add_listener()
            async for (key, notification) in self._stream.fabric.get_notifications():
                if type(notification).__name__ == StreamItemNotification.__name__:
                    cache = self._stream.ignite.get_cache(notification.cache)
                    value = cache.get(notification.key)
                    try:
                        yield self._stream.serializer.deserialize(value)
                    finally:
                        self._stream.fabric.consume_notification(key)
        finally:
            if self._stream.parameter_index < 0:
                self.__remove_listener()

    def __modify_stream_data(self, modification):
        streams_cache = self._stream.ignite.get_cache("streams")
        current_value = streams_cache.get(self._stream.stream_name)
        new_value = modification(current_value)
        # This should be replace
        streams_cache.put(self._stream.stream_name, new_value)

    def __add_listener(self):
        stream_listener = StreamListener(
            agentdelegate=self._stream.fabric.agent_delegate,
            stream=self._stream.instance.instance_name,
            parameter=self._stream.parameter_index,
            filter=(1, self._filter),
            replay=self.replay,
            localtodata=self.local_to_data,
        )

        def __add_listener_lambda_(stream_data):
            listeners = getattr(
                stream_data, "listeners", getattr(stream_data, "Listeners", None)
            )
            listeners[1].append(stream_listener)
            return stream_data

        return self.__modify_stream_data(__add_listener_lambda_)

    def __remove_listener(
        self,
    ):
        return self.__modify_stream_data(self.remove_listener_lambda)

    def __remove_listener_lambda(self, stream_data):
        current_listeners = stream_data.listeners
        bin_obj = BinaryObject()
        for listener in current_listeners:
            if (
                (listener is bin_obj)
                and (bin_obj.stream == _stream._instance.instance_name)
                and (bin_obj.parameter == _stream.kwargs["parameter_index"])
            ):
                current_listeners.remove(listener)
                break

        builder = stream_data.to_builder()
        builder.listeners = current_listeners
        return builder.build()
